### ISSUE TITLE
Spike: ChangeList

### ISSUE THREAD

**Issue by marcelgarus** [hpi-swa-teaching/ChangesBrowser#61]
- [x] Understand responsibilities of the `ChangeList` that are relevant to us
- [x] How do changes propagate in Squeak? 
- [ ] With which classes does the `ChangeList` interact?
- [x] How do `ChangeRecord`s work? What is their responsibility?
- [ ] How is a `ChangeList` supposed to be used?
- [ ] What are good abstraction layers?

Then decide:

- Re-implement the `ChangeList` or not?
- What functionality might be useful for us?
- How could we use that functionality?
- What else do we need?

Also, document insights that may be relevant to developers trying to get started with the project in the Wiki.

**Comment by marcelgarus** [comment]
> What are good abstraction layers?

Separating the model from a GUI would be a great first step.

Things to be decided:
- The name: ChangeRecordCollection? ChangeCollection?
- Exactly what responsibilities should be part of the new class?
  - Scanning the file? Which parts?
    - Byte span?
    - The number of bytes from the end?
    - By timespan?
    - By the number of days from today?
    - Only since the last snapshot/quit?
  - Holding ChangeRecords in a list?
  - Filtering? Which filters?
    - By timespan?
    - Type of change?
    - Author?
    - Concerning a specific class/method?
    - By snapshot?
    - Configuring custom filter by providing block?
- Should we support filtering when scanning?
- Destructive/non-destructive filters?
- Should we provide a new (immutable) instance for the filtered result?

**Comment by marcelgarus** [comment]
There are several types of change records:
- There's the legacy `ChangeRecord`, which is used throughout most of the `ChangeList` code.
- There's the new semi-finished `ClassChangeRecord` and `MethodChangeRecord`.
- We could write a new `ChangeRecord`.

**Comment by marcelgarus** [comment]
What we decided during our meeting:

- Name the class `ChangeRecordCollection`
- First, only scan by the number of bytes from the end. Add functionality as needed.
- We hold ChangeRecords in a list.

Filtering:
- We do not put filtering into the `ChangeRecordCollection` because of the single-responsibility principle.
- Instead, we should create a new class.

The current `ChangeRecord` is preeetty complicated. But it's used by many different classes and a rewrite would increase the scope of our project quite a bit, so we'll continue to use it.

**Comment by SilvanVerhoeven** [comment]
While implementing the selectable list items in the GUI-change-list, we noticed that we already implement some kind of `ChangeRecordCollection`, which was very close to the GUI and use-case-specific. Even worse: We instantiated a `ChangeList`, only to retrieve the `OrderedCollection` of `ChangeRecords` and delete the feature rich `ChangeList` again. That's when we developed the idea to create a `ChangeRecordLoader`, which only loads an `OrderedCollection` of `ChangeRecords` from the changes file.

For filtering, additional classes shouldn't be necessary. Instead, implementing a thoughtful and consistent property-set on the `ChangeRecord` allows the usage of the present Collection-methods like `collect` and `reject`, which seems like a cleaner approach.
